man/man1/ellipse.1
.TH ELLIPSE 1NEMO "23 October 2021"

.SH "NAME"
ellipse \- ellipse (de)projection properties

.SH "SYNOPSIS"
\fBellipse\fP [parameter=value]

.SH "DESCRIPTION"
\fBellipse\fP de-projects ellipses, and computes the intrinsic properties,
and vice versa. It also computes kinematic properties, assuming flow is
along the ellipses, such as the kinematic minor axis (the major axis
cannot be defined, since the flow-velocity along the ellipse is not
defined with this routine. 
.PP
Apart from giving the eccentricity of the ellipse ("bar") in the disk,
and the inclination of the disk, the two angles in this routine are
\fBtheta\fP and \fBphi\fP, of which only one needs to be given.
.PP
For some input parameters an array is allowed to create an output table,
which can then be easily plotted.

.SH "PARAMETERS"
The following parameters are recognized in any order if the keyword
is also given:
.TP 20
\fBba=\fP
Axis ratio of bar. An array is allowed, in \fInemoinp(1NEMO)\fP 
notation. No default.
.TP
\fBinc=\fP
Inclination of disk in which the ellipse ("bar") is located. 
An array is allowed. No default.
.TP
\fBphi=\fP
Angle between bar and disk in sky plane. An array is allowed.
Either phi or theta (not both) are required. It determines if
projection (\fBtheta\fP given) or deprojection (\fBphi\fP given) 
is done. 
.TP
\fBtheta=\fP
Angle between bar and disk in galax plane. Again an array is allowed.
.TP
\fBa=\fP
Length of the bar (currently only projected) bar. If this option is used,
the stick formulae (ba=0) will be applied, and errors can be computed
using \fBdphi=\fP.
.TP
\fBdba=\fP
Error term in b/a. 
Useful if you want to compute errors in (de)projected
parameters using nsim=
.TP
\fBdphi=\fP
Error term in phi.
.TP
\fBnsim=\fP
Number of monte carlo to perform to compute an error term. [0]
.TP
\fBseed=\fP
Seed for random number generator. [0]

.SH "EXAMPLES"
Here for a set of apparent position angles w.r.t. line of notes, and given
apparent b/a and inc, the intrinsic b/a, angle between bar and line of nodes
and the (sky plane) kinematic major axis (assumed perpendicular to the 
kinematic minor axis) are computed:
.nf

% ellipse 0.3 45 20:30:2
De-projecting ellipse:
b/a inc phi    b/a'    theta   phi_kin
0.3 45 20    0.376558 29.4085 -46.4712
0.3 45 22    0.368488 31.9773 -45.7525
0.3 45 24    0.360246 34.4675 -44.8267
0.3 45 26    0.351921 36.8805 -43.7464
0.3 45 28    0.343591 39.2183 -42.5507
0.3 45 30    0.335325 41.4834 -41.2693
.fi

and here a plot that shows how the axis ratio of a 0.5 bar changes as the bar angle
changes, for fixed inclincation of 20 degrees:
.nf
   %  ellipse 0.5 1:89:1 20 | tabplot - 2 4 line=1,1
.fi   




Here is an example of creating a bar in a disk, and viewing the sky projection. The example
is taken from NGC 253 (see also \fIrotcur(5NEMO)\fP):

.nf

# disk
  r_d=2
  pa_d=230
  inc_d=76

# bar
  a_b=1
  b_b=0.25
  pa_b=18
  
# create bar and disk
  mkconfig - 361 ring $a_b | snapscale - - rscale=$b_b,1,0 | snaprotate - bar0  $pa_b,$inc_d,$pa_d zyz
  mkconfig - 361 ring $r_d | snaprotate - disk0 $inc_d,$pa_d yz

# add, and plot the two ellipses
  snapadd disk0,bar0 - | snapplot - 

.fi


.SH "CAVEATS"
For the stick formulae (a= used) only projection is done.

.SH "SEE ALSO"
.nf
Skillman et al. A&A 198, 33 (1988), p39.
Arnaboldi et al. 1995AJ....110..199A \fIKinematics of the Ionized Gas in NGC 253 \fP
.fi

.SH "FILES"
src/orbit/misc	ellipse.c

.SH "AUTHOR"
Peter Teuben

.SH "UPDATE HISTORY"
.nf
.ta +1.0i +4.0i
fall-1986	V0.0 formulae derived for Skillman et al paper	PJT
23-Apr-03	V0.3 added phi_kin	PJT 
28-aug-03	V0.4 added a= for stick formulae of bars, +errors	PJT
.fi


src/orbit/misc/ellipse.c
/*
 *  ellipse:   (de)project ellipses at skewed angles and find the new ellipse
 *
 *   0.2  4-nov-02    original                  PJT
 *   0.3 23-apr-03    added phi_kin to try M51's oval           PJT
 *   0.4 27-aug-03    mode to compute errors in bar length from stick formulae     PJT
 *   0.4a 13-oct-2021 ansi code cleanup       PJT
 *
 */

#include <stdinc.h>
#include <getparam.h>
#include <math.h>
#include <mathfns.h>

string defv[] = {
  "ba=???\n       Axis ratio of bar",
  "inc=???\n      Inclination of galaxy",
  "phi=\n         Angle between bar and disk in sky plane ",
  "theta=\n       Angle between bar and disk in galaxy plane",
  "a=\n           Bar length (only used for stick formulae)",
  "dba=0.0\n      Error term in b/a, if determined",
  "dphi=0.0\n     Error term in phi, if determined",
  "da=0.0\n       Error term in a, if determined (only for stick formulae)",
  "nsim=0\n       Number of monte carlo to perform to compute an error term",
  "seed=0\n       Seed for random number generator",
  "VERSION=0.4a\n 13-oct-2021 PJT",
  NULL,
};

string usage = "ellipse (de)projection properties";

#define MAXP 1024
#define RPD 0.0174532925199433


void deproject(double ba, double inc, double phi, double *ba1,  double *theta, double *phi_kin) {
  double ai,bi,ci, cos2t, tan2t, ba2, tank;
  double ome2 = ba*ba;   /* 1-e^2 */
  double e2 = 1-ome2;    /* 1-b^2 ;    b := b/a, i.e.   a=1 */
  double cosp = cos(phi*RPD);
  double sinp = sin(phi*RPD);
  double cosi = cos(inc*RPD);
  double z;

  ai =           (1-e2*cosp*cosp)/ome2;
  ci = cosi*cosi*(1-e2*sinp*sinp)/ome2;
  bi = -2*e2*sinp*cosp*cosi/ome2;
  tan2t = bi/(ai-ci);
  cos2t = 1/sqrt(1+tan2t*tan2t);
  ba2 = ((ai+ci)*cos2t + (ai-ci))/((ai+ci)*cos2t - (ai-ci));
  if (ba2<0) error("Bad ba2=%g",ba2);
  z = atan(tan2t)/2/RPD;
  ba2 = sqrt(ba2);
  if (ba2 > 1) {
    ba2 = 1.0/ba2;
    z += 90.0;
  }
  *ba1 = ba2;
  *theta = z;
  tank = bi/(2*ai);
  *phi_kin = atan(tank)/RPD;
}


void  project(double ba, double inc, double theta, double *ba1,  double *phi) {
  error("project-ing not implemented yet");
}


void nemo_main()
{
  bool Qtheta = hasvalue("theta");
  bool Qphi = hasvalue("phi");
  bool Qstick = hasvalue("a");
  // bool Qerrba;
  int nsim = getiparam("nsim");
  int seed = init_xrandom(getparam("seed"));
  double ba[MAXP], inc[MAXP], ang[MAXP], x,y,z, a,a1, t1,t2,t3;
  double da;
  int nba, ninc, nang;
  int i,j,k;
  double dphi = getdparam("dphi");
  double dba = getdparam("dba");
  
  // string fmt="%g";
  dprintf(1,"seed=%d\n",seed);

  nba = nemoinpd(getparam("ba"),ba,MAXP);
  ninc = nemoinpd(getparam("inc"),inc,MAXP);

  if (Qstick) {
    /* 
     * A = a*sqrt(cosp^2+sinp^2/cosi^2)
     * a = A*sqrt(cost^2+sint^2*cosi^2)
     */
    if (hasvalue("da"))
      da = getdparam("da");
    else
      da = 0.0;
    if (Qphi) {
      dprintf(0,"Projecting stick formulae ellipse:\n");
      dprintf(0,"a   inc  theta     A  dA/A\n");
      nang = nemoinpd(getparam("phi"),ang,MAXP);
      a = getdparam("a");
      for (i=0; i<nang; i++)
	for (j=0; j<ninc; j++)
	  for (k=0; k<nba; k++) {
	    double cosi = cos(inc[j]*RPD);
	    double sini = sin(inc[j]*RPD);
	    double cosp = cos(ang[i]*RPD);
	    double sinp = sin(ang[i]*RPD);
	    x = sqrt(cosp*cosp+sinp*sinp/(cosi*cosi));
	    a1 = a*x;
	    t1 = cosp*sinp;
	    t2 = cosp*sinp/(cosi*cosi);
	    t3 = sini*sinp*sinp/(cosi*cosi*cosi);
	    y = sqrt(t1*t1 + t2*t2 + t3*t3)*dphi*RPD/(x*x);
	    if (da > 0) {
	      z = sqrt(y*y + sqr(da/a));
	    } else
	      z = y;
	    printf("%g %g %g    %g %g %g\n",a,inc[j],ang[i],a1,y,z);
	  }
    } else if (Qtheta) {
      error("theta= not supported yet");
    } else
      error("Need phi= or theta=");
  } else { 
    
    if (Qtheta) {                                    /* theta given, so project it */
      dprintf(0,"Projecting ellipse:\n");
      dprintf(0,"b/a  inc  theta     b/a'  phi\n");
      nang = nemoinpd(getparam("theta"),ang,MAXP);
      for (i=0; i<nang; i++)
	for (j=0; j<ninc; j++)
	  for (k=0; k<nba; k++) {
	    project(ba[k],inc[j],ang[i],&x,&y);
	    printf("%g %g %g    %g %g \n",ba[k],inc[j],ang[i],x,y);
	  }
    } else if (Qphi) {                               /* phi given, so deproject */
      dprintf(0,"De-projecting ellipse:\n");
      dprintf(0,"b/a  inc  phi    b/a'  theta    phi_kin\n");
      nang = nemoinpd(getparam("phi"),ang,MAXP);
      for (i=0; i<nang; i++)
	for (j=0; j<ninc; j++)
	  for (k=0; k<nba; k++) {
	    if (nsim > 0) {
	      int l;
	      for (l=0; l<nsim; l++) {
		deproject(ba[k]+grandom(0.0,dba),inc[j],ang[i]+grandom(0.0,dphi),&x,&y,&z);
		printf("%g %g %g    %g %g %g\n",ba[k],inc[j],ang[i],x,y,z);
	      }
	    } else {
	      deproject(ba[k],inc[j],ang[i],&x,&y,&z);
	      printf("%g %g %g    %g %g %g\n",ba[k],inc[j],ang[i],x,y,z);
	    }
	  }
    } else
      error("Need phi= or theta=");
  }
}

